An initial draft for the COCO architecture (#44)
confidential-containers/documentation: coco_architecture.md

These secure and isolated environments prevent unauthorized access or modification of applications and data while they are 
in use, thereby increasing the security level of organizations that manage sensitive and regulated data.

A Trusted Execution Environment (TEE) is commonly defined as an environment that provides a level of
assurance of data integrity, data confidentiality, and code integrity. A hardware-based TEE uses
hardware-backed techniques to provide increased security guarantees for the execution of code and
protection of data within that environment."

Key Characteristics of Confidential Computing (defined by Confidential Computing Consortium)

- **Data confidentiality:** Unauthorized entities cannot view data while it is in use within the TEE
- **Data integrity:** Unauthorized entities cannot add, remove, or alter data while it is in use within the TEE
- **Code integrity:** Unauthorized entities cannot add, remove, or alter code executing in the TEE

Ref: https://confidentialcomputing.io/wp-content/uploads/sites/85/2021/03/confidentialcomputing_outreach_whitepaper-8-5x11-1.pdf

# Confidential Computing (CC) Building blocks 

> ## Overview of the key components of a confidential computing solution

Author: @magowan can you review?

The following diagrams shows how the different building blocks for CC come together: 

![CC_all_blocks](https://github.com/confidential-containers/documentation/blob/main/images/CC_all_blocks.jpg)

Let’s start by describing the lower level blocks of the CC solution (**colored blue**): 

- **Infra layer** - be it on prem (bare metal, VMware etc..) or public cloud(AWS, Azure, GCP etc…)
  - **Hardware with CC support** - this includes 2 areas of technologies
  - **Process based TEEs** - In this model, a process that needs to run securely is divided into two components: trusted and untrusted. 
    The trusted component resides in encrypted memory and handles confidential computing, while the untrusted component interfaces with 
    the operating system and propagates I/O from encrypted memory to the rest of the system. Intel SGX is an example of process base TEE
- **VM based TEEs**  - In this model, memory is encrypted along a traditional VM boundary running on top of a VMM. AMD SEV-SNP, Intel TDX, 
  IBM Secure Execution and Protected Execution Functionality (PEF) are examples of VM TEEs.

> - **Hypervisor** - this includes both RH hypervisor (such as QEMU/KVM) for bare metal and non RH hypervisors from VMware or cloud providers 

Author: I've updated this to a general list of hypervisors

The **CC services block** contains a number of services (**colored pink**) which are required for creating a holistic CC platform which the customer 
can then use to build their solution. This includes the following services (we expect to see more services later on):  

- **Attestation service** -  The primary purpose of the attestation service is to validate the evidence provided by the hardware TEE. 
  This is the verifier as defined in the RATS architecture.
- **Key Broker Service (KBS)** -  The KBS is the relying party as defined in the RATS architecture.  Following are its primary function:
  - Receives evidence from the attester (confidential VM or container) via challenge-response protocol
  - Relay the evidence to the Attestation Service for verification
  - Apply appraisal policy for the returned Attestation Results to assess the trustworthiness of the attester
  - Interact with Key Management Service to retrieve the keys and send it to the attester
- **Key management service** - A service for securely storing, managing and backing up of cryptographic keys used by applications and users.

> - **Image build service** - A service that is used to build confidential container or VM images for end users

Reviewer: Building VM images is another can of worms. Not sure we want to reference that here.

Reviewer 2: Ah think i added my comment in this area to updated code. Hopefully people can correlate it to here.

- **Image registry** - A service that is used to store encrypted and/or signed container and VM images required for confidential computing workloads. 
  Examples: quay.io, docker hub, registries provided by cloud service providers etc.

On the **Confidential VM/Container side** there is this component (**colored green**)

- **Enclave software stack** - The entity responsible for collecting/sending the evidence, getting the encryption keys 
  and decrypting/running the application. This includes the attestation agent, image management capabilities etc… 
  

> ## Typical Confidential Container Stack using VM based TEEs (TDX/SEV-SNP/SE)

Reviewer: `SEV-SNP` is not yet supported. Better to just say `SEV` for now. Also I'm not totally sure that the diagram is accurate for SE. For instance the Kata Agent isn't really measured in the same way.

The following diagram shows how a CC solution is used to provide a confidential container to the end user: 

![CC_TEE_container](https://github.com/confidential-containers/documentation/blob/main/images/CC_TEE_container.jpg)

Following is the workflow when deploying a confidential container (Kubernetes POD):

- CC workload preparation

>   - User builds the container image(s)  (podman build -f …)

Reviewer: Maybe we should use `skopeo` in the example instead of podman.

Reviewer 2: Perhaps we can just saying (using tools such as skopeo of podman) since we aren't intending to give accurate low level commands as part of the architecture overview?

  - User signs/encrypts the container image(s)
  - User pushes the container image(s) to the image registry (podman push …)
- Deploying the CC workload in k8s
  - User deploys the workload (kubectl apply -f <...>)
  - Kubernetes schedules the workload to target host having the required capability to run confidential containers
- CC workload execution flow (**red connector** in the diagram)
  - Confidential containers runtime on the host starts the enclave (VM TEE)
  - Enclave (agent) performs remote attestation: **steps 1-2 from pic**
  - Enclave (agent) gets the keys required to verify/decrypt the containers image(s):  **steps 3-4 from the pic**
  - Enclave (image management) downloads the container image(s) : **step 5 from the pic**
  - Enclave verifies/decrypts the container image(s) : **step 6 from the pic**

>   - Enclave starts the container workload

Reviewer 3: Since our core components include the attestation agent, image-rs, and ocicrypt-rs it would be good to depict them here.

## Typical Confidential Container Stack using process based TEE (SGX)

The following diagram shows how a CC solution is used to provide a confidential process to the end user: 

![CC_SGX_container](https://github.com/confidential-containers/documentation/blob/main/images/CC_SGX_container.jpg)

Following is the workflow when deploying a confidential container (Kubernetes POD). 

> We have marked in blue the changes from the confidential container using VM based TEE scenario: 

Reviewer 3: I don't think the blue coloring came across from the source material

Author: Please suggest how to clarify that point.
If you find a way to color the "blue" text in blue that would also be great :-)

- CC workload preparation
  - User builds the container image(s)  (podman build -f …)
  - User signs/encrypts the container image(s)
  - User pushes the container image(s) to the image registry (podman push …)
- Deploying the CC workload in k8s
  - User deploys the workload (kubectl apply -f <...>)
  - Kubernetes schedules the workload to target host having the required capability to run confidential containers
- CC workload execution flow (**red connector** in the diagram)

>   - Confidential containers runtime on the host starts the enclave (SGX) 

Reviewer 3: ```suggestion
  - Confidential containers runtime on the host starts the enclave agent
```

  - Enclave (agent) performs remote attestation: **steps 1-2 from pic**
  - Enclave (agent) gets the keys required to verify/decrypt the containers image(s):  **steps 3-4 from the pic**
  - Enclave (image management) downloads the container image(s) : **step 5 from the pic**

>   - Enclave verifies/decrypts the container image(s) : **step 6 from the pic**

Reviewer 3: ```suggestion
  - Enclave verifies/decrypts the container image(s) and writes it to a local encrypted filesystem: **step 6 from the pic**
```

>   - Enclave starts the container workload

Reviewer 3: ```suggestion
  - The runtime starts the app enclave which reads the container bundle from the encrypted filesystem.
    - Secure use of the encrypted filesystem is facilitated by a key exchange between the agent and app enclaves using either sealing or local attestation.
```

> 

Reviewer 3: ```suggestion

The up-to-date design is maintained in the [enclave-cc documentation](https://github.com/confidential-containers/enclave-cc/blob/main/docs/design.md)

```

Author: I think we could add this in the CoCo specific part of supporting process based TEE however this section tries to explain the general approach without connecting to a specific implementation.

Please consider adding this in the last CoCo focused section.

As can be seen, the flows for process and VM based TEEs are almost identical. It should be noted that process based TEE 

> requires a few additional software components like libOS without which the application requires re-architecting. 

Reviewer 3: ```suggestion
uses a library OS as part of the App Enclave to preclude re-architecting the application. 
```

> This is not the case for VM based TEEs.

Reviewer 3: ```suggestion
This is not the case for VM based TEEs. Conversely, Process based TEEs do not require a separate VM and CC-aware hypervisor.
```

> ## Typical Confidential VM Stack

Reviewer 4: Do we want to have this section at all? A "typical" full blown confidential computing VM stack can be quite complex (OVMF, kernel, initrd, disk image, etc) and also not as well/statically defined as the confidential containers one. So instead of providing a super generic diagram, I'd suggest to just remove that section.

The following diagram shows how a CC solution is used to provide a confidential VM to the end user: 

![CC_TEE_VM](https://github.com/confidential-containers/documentation/blob/main/images/CC_TEE_VM.jpg)

Following is the workflow when deploying a confidential VM:

> - CC workload preparation

Reviewer 3: I don't know if this section is correct.
I think we provide builds for trusted guests for each vm-based TEE.
Are those installed by the Operator?
In any case, I don't think those are necessarily signed/encrypted.

  - User builds the VM image(s) 
  - User signs/encrypts the VM image(s)
  - User pushes the VM image(s) to an image registry (not shown in the diagram) 
- CC workload execution flow (**red connector** in the diagram)

>   - Confidential VM runtime on the host starts the VM TEE

Reviewer 3: Should we be more specific here.. i.e. modified containerd and kata shim?

  - Enclave (agent) performs remote attestation: **steps 1-2 from pic**
  - Enclave (agent) gets the keys:  **steps 3-4 from the pic**
  - Enclave (agent) makes the keys available to the application: **step 5**
  - The application will start running using the provided keys 

> # CNCF Confidential containers 

Reviewer 3: Cap
```suggestion
# CNCF Confidential Containers 
```

> ## Overview 

Reviewer 3: I don't understand the goal of this section. It seems to repeat the content from above but with maybe more detail. I think it would be more direct to perhaps just use these diagrams above and forego another section. Also it uses our off-hand v1 terminology which may not map to our actual release nomenclature.

Author: The goal of this section is to take the previous general explanation of how to support confidential containers with VM based TEE or process based TEE (and confidential VM stack) and to zoom into how we support it in the CoCo project using kata and enclave-CC.

Feel free to add content to help clarify things

following services:

- **Attestation service** -  The primary purpose of the attestation service is
  to validate the evidence provided by the hardware TEE. This is the *Verifier*,
  as defined in the [RATS architecture](https://datatracker.ietf.org/doc/html/draft-ietf-rats-architecture-21).
- **Key Broker Service (KBS)** -  The KBS is the *Relying Party*, as defined by
  the [RATS architecture](https://datatracker.ietf.org/doc/html/draft-ietf-rats-architecture-21).
  Following are its primary functions:
  - Receive evidence from the *Attester* (confidential VM or container) via a
    challenge-response protocol.
  - Relay the evidence to the Attestation Service for verification.
  - Apply appraisal policy for the returned Attestation Results to assess the
    trustworthiness of the *Attester*.
  - Interact with the Key Management Service to retrieve the keys and then
    send them back to the *Attester*.
- **Key management service** - A service for securely storing, managing and
  backing up of cryptographic keys used by applications and users.
- **Image build service** - A service that is used to build confidential

>   container or VM images for end users.

Reviewer 2: I worry that people may assume a single build service that builds both the confidential containers and the VM Images and that a single end user persona or role drives this build service?
 Not sure if we can change this a little? 
Perhaps "Services used to build confidential containers or VM images"

  Examples: [Quay.io](https://quay.io/), [Docker Hub](https://hub.docker.com/),
  CSPs provided registries, etc.

On the **Confidential VM/Container side** there is this component
(**colored green**)

- **Enclave software stack** - Those are trusted CoCo components responsible for
  collecting and sending the evidence, getting the encryption keys, decrypting
  the container image and running the container workload. This includes the
  attestation agent, image management capabilities, etc.

## Confidential Container Stack using a VM-based TEE (e.g. AMD SEV, IBM SE or Intel TDX)

The following diagram shows how a CC solution is used to provide a confidential
container to the end user:

![CC_TEE_container](https://github.com/confidential-containers/documentation/blob/main/images/CC_TEE_container.jpg)

Following is the workflow when deploying a confidential container (Kubernetes POD):

- CC workload preparation
  - User builds the container image(s)  (e.g. with `podman build`).
  - User signs/encrypts the container image(s).
  - User pushes the container image(s) to the image registry (e.g. with
    `podman push`).
- Deploying the CC workload in k8s

>   - User deploys the workload (`kubectl apply -f cc_workload.yaml`).

Reviewer 2: Perhaps highlight the user specifies the "CoCo" runtime in their yaml?

  - Kubernetes schedules the workload to target host having the required
    capability to run confidential containers.
- CC workload execution flow (**red connector** in the diagram)
  - Confidential containers runtime on the host starts the VM TEE (The enclave).
  - Enclave (agent) performs remote attestation: **steps 1-2 in the diagram**.
  - Enclave (agent) gets the keys required to verify/decrypt the containers
    image(s):  **steps 3-4 in the diagram**.
  - Enclave (image management) downloads the container image(s) : **step 5 in
    the diagram**.
  - Enclave verifies/decrypts the container image(s) : **step 6 in the diagram**.
  - Enclave starts the container workload.

## Confidential Container Stack using a process-based TEE (e.g. Intel SGX)

The following diagram shows how a CC solution is used to provide a confidential
process to the end user:

![CC_SGX_container](https://github.com/confidential-containers/documentation/blob/main/images/CC_SGX_container.jpg)

Following is the workflow when deploying a confidential container (Kubernetes
pod).

> We have marked in blue the changes from the confidential container using

Reviewer 2: Maybe rather than marked in blue, indicate it is the blue area of the diagram that has changed?

  - Enclave (agent) gets the keys required to verify/decrypt the containers
    image(s): **steps 3-4 in the diagram**.
  - Enclave (image management) downloads the container image(s) : **step 5 in
    the diagram**.
  - Enclave verifies, decrypts and writes the container image(s) to a local
    encrypted filesystem: **step 6 in the diagram**.
  - The runtime starts the app enclave which reads the container bundle from the
    encrypted filesystem.
  - Secure use of the encrypted filesystem is facilitated by a key exchange
    between the agent and app enclaves using either sealing or local attestation.

As can be seen, the flows for process and VM-based TEEs are almost identical. It
should be noted that process-based TEE requires a few additional software
components like `libOS` without which the application requires re-architecting.
This is not the case for VM-based TEEs. Conversely, Process based TEEs do not
require a separate VM and CC-aware hypervisor.

The up-to-date design for the Confidential Containers process-based architecture
is maintained in the [enclave-cc documentation](https://github.com/confidential-containers/enclave-cc/blob/main/docs/design.md).

> # CNCF Confidential Containers

Reviewer 3: I think this section was originally intended to be the coco specifics of generalizations made above.
That's not really the case though, as the sections above are already coco-specific. We should just move these diagrams into the section above and remove this section. The only addition in the diagrams I see is the kubelet->containerd->runtime blocks.

Alternatively the section could be demoted and called runtime and just point out that we have 2 different container runtimes.

Reviewer 2: Yes it does feel like repetition, perhaps raising up a level the CoCo area to be more abstract? Then it makes it clearer this diagram is more about the bigger picture of what is needed to use Kubernetes - so focuses on the non CoCo pieces a bit more?

Reviewer 4: I tried to re-arrange a little bit those 2 sections and make the cut between the two clearer, highlighting the fact that the Confidential Containers project is a concreter implementation of the "generic" architecture describe in the previous section. Please let me know if it looks better now.
